README.md
# Ractive Grid

A grid-like component built on [Ractive](https://github.com/ractivejs/ractive)

## Where to get it?

Racive Grid is available as a [giblet](https://github.com/evs-chris/gobble-giblet), a [component](https://github.com/componentjs/component), and a pre-assembled UMD module. Each flavor does not declare an explicit dependency on Ractive, but it is expected to be available as a global.

All of the pre-build files live in tags on the build branch.

### Development

Ractive Grid uses [gobble](https://github.com/gobblejs/gobble) as its build tool, which makes it easy to build and play around with. The default mode in the gobble file is `development`, which will automatically pull in the edge version of Ractive and make it available along with the sandbox. There is an example file provided along with the source, which you can access by running gobble and pointing you browser at http://localhost:4567/sandbox/example.html.

## Usage

Include the grid as a Ractive component and then add a component reference to it somewhere in your template. You can map the data in the grid to your base instance with the `items` attribute.

```html
<Grid items="{{myGridItems}}" />
```

By default, Grid will name your columns based on the keys of the objects in your items list. You can also supply a `columns` array containing objects with `label`, `path`, `class`, and `order` keys that specify the content of your grid columns and their header labels. You can also create button fields with a `{ type: 'button', action: function(item) { ... }, buttonLabel: 'Foo?', buttonClass: 'bar' }` single button or `{ buttons: [{ label: 'Foo?', class: 'bar', action: function(row, rowNum, colNum, btnNum, btn) { ... } }] }` multiple buttons.

Grid makes columns sortable by default. Clicking on a column header will toggle sorting for that column, and clicking again will switch the sort direction. Clicking a third time will remove any sorting. Multiple columns can participate in a sort if their headers are clicked while the `CTRL` key is held down.

Grid supplies an event handler for an input box to be used for navigation of the grid while the input is focused. The handler will cause the Grid to fire a selection event on `Enter` and will change the currently selected row with `Up` and `Down`.

## License

Copyright (c) 2014 Chris Reeves. Released under an [MIT license](https://github.com/evs-chris/ractive-window/blob/master/LICENSE.md).


CHANGELOG.md
## 0.4.0

* Adds support for button columns.
* Cleans up some Ractive migration warnings.

## 0.3.0

* Switches to ES6 via gobble and adds giblets support.
* __BUG:__ Fixes issue with setting column classes and adds support for specifying column info as a mapping.

## 0.2.0

* __BUG:__ Fixes sort-related context leak.
* Adds a helper for controlling grid selection from an external (likely filter) input field.

## 0.1.5

* __BUG:__ Avoids throwing exceptions due to missing data while the grid is initializing.
* __BUG:__ Fixes column class references.

## 0.1.4

* __BUG:__ Fixes issue with tracking current caused by attempts to fix issues with leaky context.

## 0.1.3

* __BUG:__ Fixes more (of the same?) leaky context issues by making sure that certain keypaths are local before a template is installed.

## 0.1.2

* __BUG:__ Fixes leaky context issues by switching to restricted references in templates.

## 0.1.1

* __BUG:__ Fixes column ordering in weirdo environments wherein sort functions are passed arguments in reverse. Here's lookin' at you IE and Safari.

## 0.1.0

* Adds support for div-based grids in addition to table-based grids.
* Adds support for keyboard movement within the grid.
* Adds tracking for the currently selected item and index.
* Adds support for providing a filter function and having it applied automatically.
* Adds support for passing some attributes (style, class, id) from the template component through to the actual dom element (table or div).

## 0.0.1

Initial version of simple table-based grid.


gobblefile.js
var gobble = require('gobble');
var sander = require('sander');
var path = require('path');

var parts = [], dev = gobble.env() !== 'production';

var transpilerOptions = {
  blacklist: ['modules']
};

// in dev mode, make libraries and the sandbox available
if (dev) {
  parts.push(gobble('sandbox').moveTo('sandbox'));
  parts.push(gobble('giblets').transform('giblets').moveTo('lib'));
}

var source = gobble('src');

var es5source = gobble(source).transform('6to5', transpilerOptions);

// build the main UMD module
var moduled = gobble(es5source)
  .transform('esperanto-bundle', {
    entry: 'index.js',
    dest: 'ractive-grid.js',
    type: 'umd',
    name: 'RactiveGrid',
    strict: true
  });

// for production, build the component too
if (!dev) {
  parts.push(gobble(es5source).transform('esperanto-bundle', {
    entry: 'index.js',
    dest: 'component/index.js',
    type: 'cjs',
    strict: true
  }));
}

parts.push(moduled.
  transform('relative-sourcemaps', { prefix: 'ractive-grid' })
);

// copy the assets
parts.push(gobble('assets').moveTo('assets'));

module.exports = gobble(parts);


component/index.js
'use strict';

"use strict";

/* global Ractive */

var templateTable = "<table class='ractive-grid' on-keydown='selection' tabindex='0'|>\n  <thead>\n    <tr>{{#.headers:i}}<th on-click='headerClicked:{{i}}' class='{{#(sortDir(i) === 1)}}header-sort-asc{{/}}{{#(sortDir(i) === 2)}}header-sort-desc{{/}}'>{{.}}</th>{{/}}</tr>\n  </thead>\n  <tbody>\n    {{#.rendered}}\n      {{#.itemsView:i}}\n        <tr on-click='rowClicked:{{i}}' on-dblclick='rowDoubleClicked:{{i}}' class='{{#isCurrentRow(i)}}current{{/}}'>{{>tableRow}}</tr>\n      {{/}}\n    {{/}}\n  </tbody>\n</table>";

var templateDiv = "<div class='ractive-grid' on-keydown='selection' tabindex='0'|>\n  <div class='rg-header'>{{#.headers:i}}<div on-click='headerClicked:{{i}}' class='{{#(sortDir(i) === 1)}}header-sort-asc{{/}}{{#(sortDir(i) === 2)}}header-sort-desc{{/}} {{columnClass(i)}}'>{{.}}</div>{{/}}</div>\n  <div class='rg-body'>\n    {{#.rendered}}\n      {{#.itemsView:i}}\n        <div class='rg-row{{#isCurrentRow(i)}} current{{/}}' on-dblclick='rowDoubleClicked:{{i}}' on-click='rowClicked:{{i}}' >{{>divRow}}</div>\n      {{/}}\n    {{/}}\n  </div>\n</div>";
var index__getProp = function getProp(obj, path) {
  if (!path) {
    return "";
  }var parts = path.split(".");
  for (var i = 0; i < parts.length; i++) {
    if (!!!obj) {
      return obj;
    }var p = parts[i];
    obj = obj[p];
  }
  return obj;
};

var Grid;
Grid = Ractive.extend({
  template: "{{>.type}}",
  onconstruct: function onconstruct(opts) {
    var transfer = ["class", "style", "id"];
    var str = "";
    for (var k in opts.data) if (transfer.indexOf(k) >= 0) str += " " + k + "=\"" + opts.data[k] + "\"";
    opts.partials = {};
    opts.partials.table = templateTable.replace(/\|/, str);
    opts.partials.div = templateDiv.replace(/\|/, str);
    opts.data.sorts = { order: [] };
    opts.data.type = opts.data.type || "table";
  },
  onrender: function onrender() {
    var grid = this;
    grid.on("rowDoubleClicked", function (e, i) {
      grid.fire("rowSelected", grid.get("itemsView." + i), i, e);
    });
    grid.on("rowClicked", function (e, i) {
      grid.set("currentIndex", i);
    });
    grid.on("selection", function (e) {
      var current = grid.get("currentIndex");
      if (typeof current !== "number") return;
      var items = grid.get("itemsView");
      if (e.original.keyCode === 38 && current > 0) current--;else if (e.original.keyCode === 40) current++;else if (e.original.keyCode === 13) {
        grid.fire("rowSelected", grid.get("current"), current, e);
        e.original.preventDefault();
        return;
      } else return;
      e.original.preventDefault();
      grid.current(current);
    });
    grid.on("colDoubleClicked", function (e, arr) {
      grid.fire("colSelected", e, grid.get("itemsView." + arr[0]), arr[1], arr[0]);
    });
    grid.on("headerClicked", function (e, i) {
      var sorts = grid.get("sorts");
      var state = sorts[i];
      var nstate;
      if (e.original.ctrlKey && grid.get("multisortable")) {
        nstate = 3;
        if (!!!state || state > 2) nstate = 1;else if (state === 1) nstate = 2;
        if (nstate > 0 && nstate < 3) {
          if (sorts.order.indexOf(i) < 0) sorts.order.push(i);
        }
        grid.set("sorts." + i, nstate);
      } else if (grid.get("sortable")) {
        for (var s in sorts) {
          if (s !== "order") grid.set("sorts." + s, 3);
        }
        nstate = 3;
        if (!!!state || state > 2) nstate = 1;else if (state === 1) nstate = 2;else {
          grid.set("sorts.order", []);
        }
        if (nstate > 0 && nstate < 3) grid.set("sorts.order", [i]);
        grid.set("sorts." + i, nstate);
      }
    });
  },
  data: function data() {
    return {
      rendered: false,
      editingRow: function editingRow(idx) {
        return (this.get("editingRows") || []).indexOf(idx) >= 0;
      },
      sortDir: function sortDir(idx) {
        return this.get("sorts")[idx];
      },
      columnClass: function columnClass(idx) {
        return ((this.get("_columns") || [])[idx] || {})["class"];
      },
      isCurrentRow: function isCurrentRow(idx) {
        return this.get("currentIndex") === idx;
      },
      editingRows: [],
      _columns: undefined,
      _filter: undefined,
      items: undefined,
      currentIndex: undefined,
      sorts: { order: [] },
      sortable: true,
      multisortable: true,
      table: true
    };
  },
  computed: {
    itemsView: function itemsView() {
      var grd = this;
      this.get("sorts");
      var filter = this.get("_filter");
      var sorts = this.get("sorting") || [];
      var items = (this.get("items") || []).slice(0);
      var columns = this.get("_columns") || [];
      var flted, i;
      if (typeof filter === "function") {
        flted = [];
        for (i = 0; i < items.length; i++) {
          if (filter(items[i])) flted.push(items[i]);
        }
        items = flted;
      }
      if (this.get("currentIndex") >= items.length) this.set("currentIndex", items.length - 1);
      items.sort(function (a, b) {
        for (var i in sorts) {
          var s = sorts[i];
          var dir = 1;
          var ai = index__getProp(a, columns[s.field].path);
          var bi = index__getProp(b, columns[s.field].path);

          if (s.dir === 2) dir = -1;else if (s.dir > 2) continue;

          if (ai < bi) return -1 * dir;else if (ai > bi) return 1 * dir;else continue;
        }
        return 0;
      });
      return items;
    },
    current: function current() {
      var items = this.get("itemsView");
      var idx = this.get("currentIndex");
      if (typeof idx !== "number") {
        return null;
      }return (items || [])[idx];
    },
    sorting: function sorting() {
      var res = [];
      var sorts = this.get("sorts");
      var orders = sorts.order || [];
      for (var i in orders) {
        res.push({ field: orders[i], dir: sorts[orders[i]] });
      }
      return res;
    },
    headers: function headers() {
      var cols = this.get("columnsView");
      var res = [];
      for (var i in cols) {
        var c = cols[i];
        if (!!c.label) res.push(c.label);else if (!!!c.compute) {
          var nm = c.path || "Column " + i;
          res.push(nm.substring(nm.lastIndexOf(".") + 1));
        } else res.push("Column " + i);
      }
      return res;
    },
    columnsView: function columnsView() {
      var cols = this.get("_columns");
      if (cols === undefined) cols = this.columns(this.get("columns"));
      if (cols === undefined) {
        return [];
      }cols = cols.slice(0);
      cols.sort(function (a, b) {
        var ai = a.order || 0;
        var bi = b.order || 0;
        if (ai < bi) return -1;else if (ai > bi) return 1;else return 0;
      });
      return cols;
    }
  },
  partials: {
    tableRow: "",
    divRow: "",
    table: "",
    div: ""
  },
  columns: function columns(arr) {
    var i;
    if (!arr) {
      // init columns from items
      var items = this.get("items");
      if (!!!items) {
        return [];
      }var item = items.slice(0).pop();
      var res = [];
      if (!!item) {
        var okTypes = ["String", "Number", "Boolean", "Date"];
        for (i in item) {
          if (item.hasOwnProperty(i) && !!item[i]) {
            var t = item[i].constructor.name;
            if (okTypes.indexOf(t) >= 0) res.push({ label: i, path: i });
          }
        }
      }
      return this.columns(res);
    } else {
      for (i = 0; i < arr.length; i++) if (!arr[i].hasOwnProperty("order")) arr[i].order = i;
      arr.sort(function (a, b) {
        var ai = a.order;
        var bi = b.order;
        if (ai < bi) return -1;else if (ai > bi) return 1;else return 0;
      });
      this.set("_columns", arr);
      var tstr = "",
          dstr = "";
      for (var c = 0; c < arr.length; c++) {
        tstr += "<td on-dblclick='colDoubleClicked:[{{i}}," + c + "]' class='{{columnClass(" + c + ")}}'>";
        dstr += "<div on-dblclick='colDoubleClicked:[{{i}}," + c + "]' class='{{columnClass(" + c + ")}}'>";
        if (arr[c].type === "button" || arr[c].buttons) {
          if (arr[c].buttons && Object.prototype.toString.call(arr[c].buttons) === "[object Array]") {
            for (var b = 0; b < arr[c].buttons.length; b++) {
              tstr += "<button " + (arr[c].buttons[b]["class"] ? "class=\"" + arr[c].buttons[b]["class"] + "\"" : "") + " on-click=\"clickButton(i, " + c + ", " + b + ")\">" + (arr[c].buttons[b].label || arr[c].label) + "</button>";
              dstr += "<button " + (arr[c].buttons[b]["class"] ? "class=\"" + arr[c].buttons[b]["class"] + "\"" : "") + " on-click=\"clickButton(i, " + c + ", " + b + ")\">" + (arr[c].buttons[b].label || arr[c].label) + "</button>";
            }
          } else {
            tstr += "<button " + (arr[c].buttonClass ? "class=\"" + arr[c].buttonClass + "\"" : "") + " on-click=\"clickButton(i, " + c + ")\">" + (arr[c].buttonLabel || arr[c].label) + "</button>";
            dstr += "<button " + (arr[c].buttonClass ? "class=\"" + arr[c].buttonClass + "\"" : "") + " on-click=\"clickButton(i, " + c + ")\">" + (arr[c].buttonLabel || arr[c].label) + "</button>";
          }
        } else {
          tstr += "{{" + arr[c].path + "}}";
          dstr += "{{" + arr[c].path + "}}";
        }
        tstr += "</td>";
        dstr += "</div>";
      }
      this.partials.tableRow = tstr;
      this.partials.divRow = dstr;
      this.set("rendered", false);
      this.set("rendered", true);
      return arr;
    }
  },
  current: function current(idx) {
    if (arguments.length === 0) {
      return this.get("current")();
    }var items = this.get("itemsView");
    if (idx < 0 || idx >= items.length) {
      return;
    }this.set("currentIndex", idx);
    var tbl = this.get("table");

    var n, p;
    if (tbl) {
      n = this.el.querySelector("tbody").children[idx];
      p = this.el;
    } else {
      n = this.el.querySelector(".rg-body").children[idx];
      if (n.style.overflow !== "") p = n;else p = this.el;
    }

    if (n.offsetTop < p.scrollTop) p.scrollTop = n.offsetTop;else if (n.offsetTop > p.scrollTop + p.clientHeight) p.scrollTop = n.offsetTop - p.clientHeight / 2;
  },
  filter: function filter(fn) {
    this.set("_filter", fn);
    this.update("itemsView");
  },
  filterInputControl: function filterInputControl() {
    var grid = this;
    return function (e) {
      var k = e.original.keyCode;
      if (k === 13) {
        var cur = grid.get("current");
        if (!!cur) {
          grid.fire("rowSelected", cur, null, null);
          e.original.preventDefault();
          return;
        }
      }
      if (k === 38 || k === 40) {
        e.original.preventDefault();
        var current = grid.get("currentIndex");
        if (typeof current !== "number") current = k === 38 ? grid.get("itemsView").length - 1 : 0;else current = k == 38 ? current - 1 : current + 1;
        grid.current(current);
        return;
      }
    };
  },
  clickButton: function clickButton(rown, coln, btnn) {
    var col = this.get("_columns." + coln),
        row = this.get("itemsView." + rown);
    if (col.buttons && col.buttons[btnn]) {
      return col.buttons[btnn].action.call(this, row, rown, coln, btnn, col.buttons[btnn]);
    } else {
      return col.action.call(this, row, rown, coln);
    }
  }
});

var index = Grid;

exports['default'] = index;
//# sourceMappingURL=index.js.map


ractive-grid.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  factory((global.RactiveGrid = {}))
}(this, function (exports) { 'use strict';

  "use strict";

  /* global Ractive */

  var templateTable = "<table class='ractive-grid' on-keydown='selection' tabindex='0'|>\n  <thead>\n    <tr>{{#.headers:i}}<th on-click='headerClicked:{{i}}' class='{{#(sortDir(i) === 1)}}header-sort-asc{{/}}{{#(sortDir(i) === 2)}}header-sort-desc{{/}}'>{{.}}</th>{{/}}</tr>\n  </thead>\n  <tbody>\n    {{#.rendered}}\n      {{#.itemsView:i}}\n        <tr on-click='rowClicked:{{i}}' on-dblclick='rowDoubleClicked:{{i}}' class='{{#isCurrentRow(i)}}current{{/}}'>{{>tableRow}}</tr>\n      {{/}}\n    {{/}}\n  </tbody>\n</table>";

  var templateDiv = "<div class='ractive-grid' on-keydown='selection' tabindex='0'|>\n  <div class='rg-header'>{{#.headers:i}}<div on-click='headerClicked:{{i}}' class='{{#(sortDir(i) === 1)}}header-sort-asc{{/}}{{#(sortDir(i) === 2)}}header-sort-desc{{/}} {{columnClass(i)}}'>{{.}}</div>{{/}}</div>\n  <div class='rg-body'>\n    {{#.rendered}}\n      {{#.itemsView:i}}\n        <div class='rg-row{{#isCurrentRow(i)}} current{{/}}' on-dblclick='rowDoubleClicked:{{i}}' on-click='rowClicked:{{i}}' >{{>divRow}}</div>\n      {{/}}\n    {{/}}\n  </div>\n</div>";
  var index__getProp = function getProp(obj, path) {
    if (!path) {
      return "";
    }var parts = path.split(".");
    for (var i = 0; i < parts.length; i++) {
      if (!!!obj) {
        return obj;
      }var p = parts[i];
      obj = obj[p];
    }
    return obj;
  };

  var Grid;
  Grid = Ractive.extend({
    template: "{{>.type}}",
    onconstruct: function onconstruct(opts) {
      var transfer = ["class", "style", "id"];
      var str = "";
      for (var k in opts.data) if (transfer.indexOf(k) >= 0) str += " " + k + "=\"" + opts.data[k] + "\"";
      opts.partials = {};
      opts.partials.table = templateTable.replace(/\|/, str);
      opts.partials.div = templateDiv.replace(/\|/, str);
      opts.data.sorts = { order: [] };
      opts.data.type = opts.data.type || "table";
    },
    onrender: function onrender() {
      var grid = this;
      grid.on("rowDoubleClicked", function (e, i) {
        grid.fire("rowSelected", grid.get("itemsView." + i), i, e);
      });
      grid.on("rowClicked", function (e, i) {
        grid.set("currentIndex", i);
      });
      grid.on("selection", function (e) {
        var current = grid.get("currentIndex");
        if (typeof current !== "number") return;
        var items = grid.get("itemsView");
        if (e.original.keyCode === 38 && current > 0) current--;else if (e.original.keyCode === 40) current++;else if (e.original.keyCode === 13) {
          grid.fire("rowSelected", grid.get("current"), current, e);
          e.original.preventDefault();
          return;
        } else return;
        e.original.preventDefault();
        grid.current(current);
      });
      grid.on("colDoubleClicked", function (e, arr) {
        grid.fire("colSelected", e, grid.get("itemsView." + arr[0]), arr[1], arr[0]);
      });
      grid.on("headerClicked", function (e, i) {
        var sorts = grid.get("sorts");
        var state = sorts[i];
        var nstate;
        if (e.original.ctrlKey && grid.get("multisortable")) {
          nstate = 3;
          if (!!!state || state > 2) nstate = 1;else if (state === 1) nstate = 2;
          if (nstate > 0 && nstate < 3) {
            if (sorts.order.indexOf(i) < 0) sorts.order.push(i);
          }
          grid.set("sorts." + i, nstate);
        } else if (grid.get("sortable")) {
          for (var s in sorts) {
            if (s !== "order") grid.set("sorts." + s, 3);
          }
          nstate = 3;
          if (!!!state || state > 2) nstate = 1;else if (state === 1) nstate = 2;else {
            grid.set("sorts.order", []);
          }
          if (nstate > 0 && nstate < 3) grid.set("sorts.order", [i]);
          grid.set("sorts." + i, nstate);
        }
      });
    },
    data: function data() {
      return {
        rendered: false,
        editingRow: function editingRow(idx) {
          return (this.get("editingRows") || []).indexOf(idx) >= 0;
        },
        sortDir: function sortDir(idx) {
          return this.get("sorts")[idx];
        },
        columnClass: function columnClass(idx) {
          return ((this.get("_columns") || [])[idx] || {})["class"];
        },
        isCurrentRow: function isCurrentRow(idx) {
          return this.get("currentIndex") === idx;
        },
        editingRows: [],
        _columns: undefined,
        _filter: undefined,
        items: undefined,
        currentIndex: undefined,
        sorts: { order: [] },
        sortable: true,
        multisortable: true,
        table: true
      };
    },
    computed: {
      itemsView: function itemsView() {
        var grd = this;
        this.get("sorts");
        var filter = this.get("_filter");
        var sorts = this.get("sorting") || [];
        var items = (this.get("items") || []).slice(0);
        var columns = this.get("_columns") || [];
        var flted, i;
        if (typeof filter === "function") {
          flted = [];
          for (i = 0; i < items.length; i++) {
            if (filter(items[i])) flted.push(items[i]);
          }
          items = flted;
        }
        if (this.get("currentIndex") >= items.length) this.set("currentIndex", items.length - 1);
        items.sort(function (a, b) {
          for (var i in sorts) {
            var s = sorts[i];
            var dir = 1;
            var ai = index__getProp(a, columns[s.field].path);
            var bi = index__getProp(b, columns[s.field].path);

            if (s.dir === 2) dir = -1;else if (s.dir > 2) continue;

            if (ai < bi) return -1 * dir;else if (ai > bi) return 1 * dir;else continue;
          }
          return 0;
        });
        return items;
      },
      current: function current() {
        var items = this.get("itemsView");
        var idx = this.get("currentIndex");
        if (typeof idx !== "number") {
          return null;
        }return (items || [])[idx];
      },
      sorting: function sorting() {
        var res = [];
        var sorts = this.get("sorts");
        var orders = sorts.order || [];
        for (var i in orders) {
          res.push({ field: orders[i], dir: sorts[orders[i]] });
        }
        return res;
      },
      headers: function headers() {
        var cols = this.get("columnsView");
        var res = [];
        for (var i in cols) {
          var c = cols[i];
          if (!!c.label) res.push(c.label);else if (!!!c.compute) {
            var nm = c.path || "Column " + i;
            res.push(nm.substring(nm.lastIndexOf(".") + 1));
          } else res.push("Column " + i);
        }
        return res;
      },
      columnsView: function columnsView() {
        var cols = this.get("_columns");
        if (cols === undefined) cols = this.columns(this.get("columns"));
        if (cols === undefined) {
          return [];
        }cols = cols.slice(0);
        cols.sort(function (a, b) {
          var ai = a.order || 0;
          var bi = b.order || 0;
          if (ai < bi) return -1;else if (ai > bi) return 1;else return 0;
        });
        return cols;
      }
    },
    partials: {
      tableRow: "",
      divRow: "",
      table: "",
      div: ""
    },
    columns: function columns(arr) {
      var i;
      if (!arr) {
        // init columns from items
        var items = this.get("items");
        if (!!!items) {
          return [];
        }var item = items.slice(0).pop();
        var res = [];
        if (!!item) {
          var okTypes = ["String", "Number", "Boolean", "Date"];
          for (i in item) {
            if (item.hasOwnProperty(i) && !!item[i]) {
              var t = item[i].constructor.name;
              if (okTypes.indexOf(t) >= 0) res.push({ label: i, path: i });
            }
          }
        }
        return this.columns(res);
      } else {
        for (i = 0; i < arr.length; i++) if (!arr[i].hasOwnProperty("order")) arr[i].order = i;
        arr.sort(function (a, b) {
          var ai = a.order;
          var bi = b.order;
          if (ai < bi) return -1;else if (ai > bi) return 1;else return 0;
        });
        this.set("_columns", arr);
        var tstr = "",
            dstr = "";
        for (var c = 0; c < arr.length; c++) {
          tstr += "<td on-dblclick='colDoubleClicked:[{{i}}," + c + "]' class='{{columnClass(" + c + ")}}'>";
          dstr += "<div on-dblclick='colDoubleClicked:[{{i}}," + c + "]' class='{{columnClass(" + c + ")}}'>";
          if (arr[c].type === "button" || arr[c].buttons) {
            if (arr[c].buttons && Object.prototype.toString.call(arr[c].buttons) === "[object Array]") {
              for (var b = 0; b < arr[c].buttons.length; b++) {
                tstr += "<button " + (arr[c].buttons[b]["class"] ? "class=\"" + arr[c].buttons[b]["class"] + "\"" : "") + " on-click=\"clickButton(i, " + c + ", " + b + ")\">" + (arr[c].buttons[b].label || arr[c].label) + "</button>";
                dstr += "<button " + (arr[c].buttons[b]["class"] ? "class=\"" + arr[c].buttons[b]["class"] + "\"" : "") + " on-click=\"clickButton(i, " + c + ", " + b + ")\">" + (arr[c].buttons[b].label || arr[c].label) + "</button>";
              }
            } else {
              tstr += "<button " + (arr[c].buttonClass ? "class=\"" + arr[c].buttonClass + "\"" : "") + " on-click=\"clickButton(i, " + c + ")\">" + (arr[c].buttonLabel || arr[c].label) + "</button>";
              dstr += "<button " + (arr[c].buttonClass ? "class=\"" + arr[c].buttonClass + "\"" : "") + " on-click=\"clickButton(i, " + c + ")\">" + (arr[c].buttonLabel || arr[c].label) + "</button>";
            }
          } else {
            tstr += "{{" + arr[c].path + "}}";
            dstr += "{{" + arr[c].path + "}}";
          }
          tstr += "</td>";
          dstr += "</div>";
        }
        this.partials.tableRow = tstr;
        this.partials.divRow = dstr;
        this.set("rendered", false);
        this.set("rendered", true);
        return arr;
      }
    },
    current: function current(idx) {
      if (arguments.length === 0) {
        return this.get("current")();
      }var items = this.get("itemsView");
      if (idx < 0 || idx >= items.length) {
        return;
      }this.set("currentIndex", idx);
      var tbl = this.get("table");

      var n, p;
      if (tbl) {
        n = this.el.querySelector("tbody").children[idx];
        p = this.el;
      } else {
        n = this.el.querySelector(".rg-body").children[idx];
        if (n.style.overflow !== "") p = n;else p = this.el;
      }

      if (n.offsetTop < p.scrollTop) p.scrollTop = n.offsetTop;else if (n.offsetTop > p.scrollTop + p.clientHeight) p.scrollTop = n.offsetTop - p.clientHeight / 2;
    },
    filter: function filter(fn) {
      this.set("_filter", fn);
      this.update("itemsView");
    },
    filterInputControl: function filterInputControl() {
      var grid = this;
      return function (e) {
        var k = e.original.keyCode;
        if (k === 13) {
          var cur = grid.get("current");
          if (!!cur) {
            grid.fire("rowSelected", cur, null, null);
            e.original.preventDefault();
            return;
          }
        }
        if (k === 38 || k === 40) {
          e.original.preventDefault();
          var current = grid.get("currentIndex");
          if (typeof current !== "number") current = k === 38 ? grid.get("itemsView").length - 1 : 0;else current = k == 38 ? current - 1 : current + 1;
          grid.current(current);
          return;
        }
      };
    },
    clickButton: function clickButton(rown, coln, btnn) {
      var col = this.get("_columns." + coln),
          row = this.get("itemsView." + rown);
      if (col.buttons && col.buttons[btnn]) {
        return col.buttons[btnn].action.call(this, row, rown, coln, btnn, col.buttons[btnn]);
      } else {
        return col.action.call(this, row, rown, coln);
      }
    }
  });

  var index = Grid;

  exports['default'] = index;

}));
//# sourceMappingURL=ractive-grid.js.map
